### DISCUSSION TITLE
[ADKernel][Custom Kernel Problem][adCoupledGradient error]

### DISCUSSION THREAD

**Discussion by coskrrb2002** [idaholab/moose#16406] [Q&A Modules: General] [answered]
Dear Moose users,

I am trying to make a simple ADKernel in moose application.

But I have a problem with adCoupledGradient term!

When I am trying to compile it, the following error has pop up.
-------------------------------------------------------------------
In file included from /opt/moose/moose/scripts/../libmesh/installed/include/metaphysicl/dualnumberarray.h:32:0,
                 from /opt/moose/moose/framework/build/header_symlinks/NumberArrayOps.h:14,
                 from /opt/moose/moose/framework/build/header_symlinks/DualRealOps.h:18,
                 from /opt/moose/moose/framework/build/header_symlinks/ADReal.h:13,
                 from /opt/moose/moose/framework/build/header_symlinks/MooseTypes.h:13,
                 from /opt/moose/moose/framework/build/header_symlinks/InputParameters.h:15,
                 from /opt/moose/moose/framework/build/header_symlinks/MooseObject.h:13,
                 from /opt/moose/moose/framework/build/header_symlinks/ResidualObject.h:12,
                 from /opt/moose/moose/framework/build/header_symlinks/KernelBase.h:12,
                 from /opt/moose/moose/framework/build/header_symlinks/ADKernel.h:12,
                 from /home/crud_kaist1/projects/CCA-E/build/header_symlinks/ChargeT.h:3,
                 from /home/crud_kaist1/projects/CCA-E/src/kernels/ChargeT.C:1,
                 from /home/crud_kaist1/projects/CCA-E/build/unity_src/kernels_Unity.C:2:
/opt/moose/moose/scripts/../libmesh/installed/include/metaphysicl/dualnumber_decl.h: In instantiation of ‘static T MetaPhysicL::DualNumberConstructor<T, D, asd>::value(const T2&) [with T2 = libMesh::TypeVector<MetaPhysicL::DualNumber<double, MetaPhysicL::NumberArray<50ul, double>, true> >; T = double; D = MetaPhysicL::NumberArray<50ul, double>; bool asd = true]’:
/opt/moose/moose/scripts/../libmesh/installed/include/metaphysicl/dualnumber.h:202:53:   required from ‘MetaPhysicL::DualNumber<T, D, asd>::DualNumber(const T2&) [with T2 = libMesh::TypeVector<MetaPhysicL::DualNumber<double, MetaPhysicL::NumberArray<50ul, double>, true> >; T = double; D = MetaPhysicL::NumberArray<50ul, double>; bool asd = true]’
/home/crud_kaist1/projects/CCA-E/src/kernels/ChargeT.C:39:85:   required from here
/opt/moose/moose/scripts/../libmesh/installed/include/metaphysicl/dualnumber_decl.h:170:40: error: cannot convert ‘const libMesh::TypeVector<MetaPhysicL::DualNumber<double, MetaPhysicL::NumberArray<50ul, double>, true> >’ to ‘double’ in return
   static T value(const T2& v) { return v; }
                                        ^
/opt/moose/moose/scripts/../libmesh/installed/include/metaphysicl/dualnumber_decl.h: In static member function ‘static T MetaPhysicL::DualNumberConstructor<T, D, asd>::value(const T2&) [with T2 = libMesh::TypeVector<MetaPhysicL::DualNumber<double, MetaPhysicL::NumberArray<50ul, double>, true> >; T = double; D = MetaPhysicL::NumberArray<50ul, double>; bool asd = true]’:
/opt/moose/moose/scripts/../libmesh/installed/include/metaphysicl/dualnumber_decl.h:170:43: error: control reaches end of non-void function [-Werror=return-type]
   static T value(const T2& v) { return v; }
                                           ^
cc1plus: some warnings being treated as errors
make: *** [/home/crud_kaist1/projects/CCA-E/build/unity_src/kernels_Unity.x86_64-pc-linux-gnu.opt.lo] Error 1

-------------------------------------------------------------------
I can't understand what's this error...

How can I fix this error?

If I eliminate adCoupledGradient terms, the kernel works fine.

The following is .c and .h file of my custom ADKernel.

------------------------------------------------------------------  
#pragma once

#include "ADKernel.h"

class ChargeT;

template <>
InputParameters validParams<ChargeT>();

class ChargeT : public ADKernel
{
public:
 static InputParameters validParams();

 ChargeT(const InputParameters & parameters);

protected:
 virtual ADReal computeQpResidual();

 const ADVariableValue & _EP;
 const ADVariableValue & _T;

 Real _z;
 Real _D;

 const Real _F;
 const Real _R;

 const ADVariableGradient & _grad_EP;

};
------------------------------------------------------------------

#include "ChargeT.h"
#include "MooseVariable.h"

registerMooseObject("corrosionApp",ChargeT);

template<>
InputParameters
validParams<ChargeT>()
{
  InputParameters params = validParams<ADKernel>();
  params.addClassDescription("Compute the migration term.");
  params.addRequiredCoupledVar("Potential", "Potential of solution");
  params.addRequiredCoupledVar("Temperature", "Temperature of solution");

  params.addRequiredParam<Real>("Charge", "Charge of chemical species");
  params.addRequiredParam<Real>("Diffusion_coef", "Charge of chemical species");

  return params;
}

ChargeT::ChargeT(const InputParameters & parameters)
  : ADKernel(parameters),
    _EP(adCoupledValue("Potential")),
    _T(adCoupledValue("Temperature")),
    _z(getParam<Real>("Charge")),
    _D(getParam<Real>("diffusion_coef")),
    _F(964800),
    _R(8.314),

    _grad_EP(adCoupledGradient("Potential"))

{
}

ADReal
ChargeT::computeQpResidual()
{
  return -_test[_i][_qp] * _D * _z * _F / ( _R * _T[_qp] ) * _u[_qp]  * _grad_EP[_qp];
}

**Comment by coskrrb2002** [comment]
Never mind!

I found my mistake in here

**Reply by aeslaughter** [reply]
Can you please post the mistake so others searching the forum can see the answer?

**Comment by GiudGiud** [comment, marked as answer]
was probably getParam and addParam missing their template specialization
